enterprise/tailnet/pgcoord_test.go

```diff
@@ -913,6 +929,287 @@ func TestPGCoordinatorDual_PeerReconnect(t *testing.T) {
+					Pubsub:           ps,
+					DeploymentValues: dv,
+				},
+			})
+			client2.SetSessionToken(client1.SessionToken())
+
+			authToken := uuid.NewString()
+			version := coderdtest.CreateTemplateVersion(t, client1, user.OrganizationID, &echo.Responses{
+				Parse:          echo.ParseComplete,
+				ProvisionPlan:  echo.PlanComplete,
+				ProvisionApply: echo.ProvisionApplyWithAgent(authToken),
+			})
+			template := coderdtest.CreateTemplate(t, client1, user.OrganizationID, version.ID)
+			_ = coderdtest.AwaitTemplateVersionJobCompleted(t, client1, version.ID)
+			workspace := coderdtest.CreateWorkspace(t, client1, template.ID)
+			_ = coderdtest.AwaitWorkspaceBuildJobCompleted(t, client1, workspace.LatestBuild.ID)
+
+			// Agent connects via the first coordinator.
+			_ = agenttest.New(t, client1.URL, authToken, func(o *agent.Options) {
+				o.Logger = logger.Named("agent1")
+			})
+			resources := coderdtest.NewWorkspaceAgentWaiter(t, client1, workspace.ID).Wait()
+
+			agentID := uuid.Nil
+			for _, r := range resources {
+				for _, a := range r.Agents {
+					agentID = a.ID
+					break
+				}
+			}
+			require.NotEqual(t, uuid.Nil, agentID)
+
+			// Client connects via the second coordinator.
+			ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitSuperLong)
+			defer cancel()
+			workspaceClient2 := workspacesdk.New(client2)
+			conn, err := workspaceClient2.DialAgent(ctx, agentID, &workspacesdk.DialAgentOptions{
+				Logger: logger.Named("client"),
+			})
+			require.NoError(t, err)
+			defer conn.Close()
+
+			_, p2p, _, err := conn.Ping(ctx)
+			require.NoError(t, err)
+			require.Equal(t, direct, p2p, "mismatched p2p state")
+
+			// Open a TCP server and connection to it through the tunnel that
+			// should be maintained throughout the restart.
+			tcpServerAddr := tcpEchoServer(t)
+			tcpConn, err := conn.DialContext(ctx, "tcp", tcpServerAddr)
+			require.NoError(t, err)
+			defer tcpConn.Close()
+			writeReadEcho(t, ctx, tcpConn)
+
+			// Stop the first server.
+			logger.Info(ctx, "test: stopping server 1")
+			s1.Stop(t)
+
+			// Pings should fail on DERP but succeed on direct connections.
+			pingCtx, pingCancel := context.WithTimeout(ctx, 2*time.Second) //nolint:gocritic // it's going to hang and timeout for DERP, so this needs to be short
+			defer pingCancel()
+			_, p2p, _, err = conn.Ping(pingCtx)
+			if direct {
+				require.NoError(t, err)
+				require.True(t, p2p, "expected direct connection")
+			} else {
+				require.ErrorIs(t, err, context.DeadlineExceeded)
+			}
+
+			// The existing TCP connection should still be working if we're
+			// using direct connections.
+			if direct {
+				writeReadEcho(t, ctx, tcpConn)
+			}
+
+			// Start the first server again.
+			logger.Info(ctx, "test: starting server 1")
+			s1.Start(t)
+
+			// Restart the second server.
+			logger.Info(ctx, "test: stopping server 2")
+			s2.Stop(t)
+			logger.Info(ctx, "test: starting server 2")
+			s2.Start(t)
```

Comment:
On restart, I'm seeing server2 fail to connect to its internal DERP server in a loop

```
    t.go:99: 2024-10-22 09:48:36.263 [debu]  server2.servertailnet: connecting to embedded DERP via in-memory pipe
    t.go:99: 2024-10-22 09:48:36.264 [debu]  server2.derp: derp: internal: client [DDvh4]: readFrameHeader: io: read/write on closed pipe
    t.go:99: 2024-10-22 09:48:36.264 [debu]  server2.servertailnet.net.wgengine: magicsock: derp-999 connected; connGen=183
    t.go:99: 2024-10-22 09:48:36.264 [debu]  server2.servertailnet.net.wgengine: magicsock: [0x14001628000] derp.Recv(derp-999): derp.Recv: io: read/write on closed pipe
    t.go:99: 2024-10-22 09:48:36.264 [debu]  server2.servertailnet.net.wgengine: derp-999: [v1] backoff: 14 msec
    t.go:99: 2024-10-22 09:48:36.279 [debu]  server2.servertailnet: connecting to embedded DERP via in-memory pipe
    t.go:99: 2024-10-22 09:48:36.279 [debu]  server2.derp: derp: internal: client [DDvh4]: readFrameHeader: io: read/write on closed pipe
    t.go:99: 2024-10-22 09:48:36.279 [debu]  server2.servertailnet.net.wgengine: magicsock: derp-999 connected; connGen=184
    t.go:99: 2024-10-22 09:48:36.279 [debu]  server2.servertailnet.net.wgengine: magicsock: [0x14001628000] derp.Recv(derp-999): derp.Recv: io: read/write on closed pipe
    t.go:99: 2024-10-22 09:48:36.279 [debu]  server2.servertailnet.net.wgengine: derp-999: [v1] backoff: 10 msec
```

The test passes, but something is fishy about the way we are restarting the server.

Reply:
I was able to reproduce the same problem by just stopping a server with an existing agent connection. The log spam lasts the whole 10 seconds as it waits for all the active websocket connections to be closed. Not sure why the magicsock would be closing the `net.Conn` it gets from the dialer, but I guess that's whats happening, but only during shutdown.

```
conn.SetDERPRegionDialer(func(_ context.Context, region *tailcfg.DERPRegion) net.Conn {
	if !region.EmbeddedRelay {
		return nil
	}
	logger.Debug(ctx, "connecting to embedded DERP via in-memory pipe")
	left, right := net.Pipe()
	go func() {
		defer left.Close()
		defer right.Close()
		brw := bufio.NewReadWriter(bufio.NewReader(right), bufio.NewWriter(right))
		derpServer.Accept(ctx, right, brw, "internal")
	}()
	return left
})
```

Reply:
I don't think it's the `magicsock` --- in the case of `coderd`'s connection to DERP, we use an in-memory pipe to connect to the local DERP server.  That pipe is somehow getting closed.

Reply:
Figured it out - `derpServer.Accept` closes the passed Conn if the ctx it's given is cancelled. There's a nil check on the output of the callback in the DERP HTTP client, so we can just rewrite it to return nil if we're shutting down.

```
		conn.SetDERPRegionDialer(func(_ context.Context, region *tailcfg.DERPRegion) net.Conn {
			if !region.EmbeddedRelay {
				return nil
			}
			select {
			case <-ctx.Done():
				return nil
			default:
			}
			logger.Debug(ctx, "connecting to embedded DERP via in-memory pipe")
			left, right := net.Pipe()
			go func() {
				defer left.Close()
				defer right.Close()
				brw := bufio.NewReadWriter(bufio.NewReader(right), bufio.NewWriter(right))
				derpServer.Accept(ctx, right, brw, "internal")
			}()
			return left
		})
```